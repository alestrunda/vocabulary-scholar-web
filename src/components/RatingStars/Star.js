import React from 'react';
import PropTypes from 'prop-types';
import ToggleStar from 'material-ui/svg-icons/toggle/star';

const RatingStar = props => {
  const handleClick = () => {
    props.onClick && props.onClick(props.number);
  };

  const handleMouseEnter = () => {
    props.onMouseEnter && props.onMouseEnter(props.number);
  };

  const handleMouseLeave = () => {
    props.onMouseLeave && props.onMouseLeave(props.number);
  };

  return (
    <div
      onClick={handleClick}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      className={`rating-stars__star rating-stars__star--${
        props.active ? 'active' : 'unactive'
      }`}
      style={props.style}
    >
      <ToggleStar
        className="rating-stars__icon"
        color={props.color}
        style={{
          width: props.size,
          height: props.size,
          ...props.styleStar,
        }}
      />
    </div>
  );
};

RatingStar.propTypes = {
  number: PropTypes.number,
  size: PropTypes.number,
  color: PropTypes.string,
  onClick: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  style: PropTypes.object,
  styleStar: PropTypes.object,
  active: PropTypes.bool,
};

RatingStar.defaultProps = {
  active: false,
};

export default RatingStar;
