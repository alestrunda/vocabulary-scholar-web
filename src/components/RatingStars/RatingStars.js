import React from 'react';
import PropTypes from 'prop-types';
import Star from './Star';
import classNames from 'classnames';

class RatingStars extends React.Component {
  state = {
    starNumberHovered: 0,
  };

  handleStarClick = starNumber => {
    this.props.onStarClick && this.props.onStarClick(starNumber);
  };

  handleStarMouseEnter = starNumber => {
    this.setState({
      starNumberHovered: starNumber,
    });
  };

  handleStarMouseLeave = () => {
    this.setState({
      starNumberHovered: 0,
    });
  };

  getStarColor = (starNumber, isActiveStar) => {
    const normalColor = isActiveStar
      ? this.props.colorActive
      : this.props.colorUnactive;
    const hoverColor = this.props.hoverColor;
    return this.props.hoverColor && this.state.starNumberHovered >= starNumber
      ? hoverColor
      : normalColor;
  };

  renderStar(number, isActive) {
    const testid = isActive ? 'rating-star-active' : 'rating-star-unactive';
    const color = this.getStarColor(number, isActive);
    return (
      <Star
        data-testid={testid}
        onMouseLeave={this.handleStarMouseLeave}
        onMouseEnter={this.handleStarMouseEnter}
        onClick={this.handleStarClick}
        key={number}
        size={this.props.iconSize}
        number={number}
        color={color}
        active={isActive}
      />
    );
  }

  render() {
    const starsActiveCount = Math.round(this.props.stars);

    //create active stars
    let isStarActive = true;
    let starsActive;
    if (this.props.stars === 0) {
      starsActive = [];
    } else {
      starsActive = Array.apply(null, Array(starsActiveCount)); //init empty array for the active stars
      starsActive = starsActive.map((item, index) => {
        const starNumber = index + 1;
        return this.renderStar(starNumber, isStarActive);
      });
    }

    //create unactive stars
    isStarActive = false;
    let starsUnactive = Array.apply(
      //init empty array for the unactive stars
      null,
      Array(this.props.length - starsActiveCount)
    );
    starsUnactive = starsUnactive.map((item, index) => {
      const starNumber = starsActiveCount + index + 1;
      return this.renderStar(starNumber, isStarActive);
    });

    return (
      <div
        className={classNames('rating-stars', this.props.className)}
        style={{
          width: this.props.iconSize * this.props.length,
          height: this.props.iconSize,
          lineHeight: `${this.props.iconSize}px`,
          ...this.props.style,
        }}
      >
        <div className="rating-stars__active">{starsActive}</div>
        <div className="rating-stars__unactive">{starsUnactive}</div>
      </div>
    );
  }
}

RatingStars.propTypes = {
  className: PropTypes.string,
  length: PropTypes.number,
  stars: PropTypes.number,
  iconSize: PropTypes.number,
  colorActive: PropTypes.string,
  colorUnactive: PropTypes.string,
  onStarClick: PropTypes.func,
  style: PropTypes.object,
  hoverColor: PropTypes.string,
};

RatingStars.defaultProps = {
  length: 5,
  stars: 0,
  iconSize: 15,
  colorActive: '#00adb5',
  colorUnactive: '#ccc',
};

export default RatingStars;
