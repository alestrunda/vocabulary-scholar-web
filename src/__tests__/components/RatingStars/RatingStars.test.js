import React from 'react';
import { shallow, mount } from 'enzyme';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

import RatingStars from '../../../components/RatingStars';

it('renders correctly', () => {
  const ratingStars = shallow(<RatingStars stars={2} />);
  expect(ratingStars).toMatchSnapshot();
});

it('renders stars', () => {
  const length = 6,
    activeStars = 3,
    iconSize = 20;
  const ratingStars = shallow(
    <RatingStars iconSize={iconSize} length={length} stars={activeStars} />
  );
  expect(ratingStars.prop('style')).toHaveProperty('width', iconSize * length);
  expect(ratingStars.find('[data-testid="rating-star-unactive"]').length).toBe(
    length - activeStars
  );
  expect(ratingStars.find('[data-testid="rating-star-active"]').length).toBe(
    activeStars
  );
});

it('triggers onStarClick when Star clicked', () => {
  const activeStars = 2,
    onClick = jest.fn();
  const root = mount(
    <MuiThemeProvider>
      <RatingStars onStarClick={onClick} stars={activeStars} />
    </MuiThemeProvider>
  );
  const ratingStars = root.find('RatingStars');
  ratingStars
    .find('[data-testid="rating-star-unactive"]')
    .first()
    .simulate('click');
  expect(onClick).toHaveBeenCalledWith(activeStars + 1);
});

it('colors correct stars when Star hovered', () => {
  const activeStars = 3,
    colorUnactive = '#000',
    hoverColor = '#f00';
  const root = mount(
    <MuiThemeProvider>
      <RatingStars
        colorUnactive={colorUnactive}
        hoverColor={hoverColor}
        stars={activeStars}
      />
    </MuiThemeProvider>
  );
  const ratingStars = root.find('RatingStars');
  ratingStars
    .find('[data-testid="rating-star-unactive"]')
    .first()
    .simulate('mouseenter');
  ratingStars.find('Star').map(item => {
    if (item > activeStars) expect(item.prop('color')).toBe(colorUnactive);
    else expect(item.prop('color')).toBe(hoverColor);
  });
});
