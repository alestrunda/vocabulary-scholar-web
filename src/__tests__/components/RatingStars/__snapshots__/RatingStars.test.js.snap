// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RatingStars
    colorActive="#00adb5"
    colorUnactive="#ccc"
    iconSize={15}
    length={5}
    stars={2}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="rating-stars__active"
        >
          <RatingStar
            active={true}
            color="#00adb5"
            data-testid="rating-star-active"
            number={1}
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            size={15}
          />
          <RatingStar
            active={true}
            color="#00adb5"
            data-testid="rating-star-active"
            number={2}
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            size={15}
          />
        </div>,
        <div
          className="rating-stars__unactive"
        >
          <RatingStar
            active={false}
            color="#ccc"
            data-testid="rating-star-unactive"
            number={3}
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            size={15}
          />
          <RatingStar
            active={false}
            color="#ccc"
            data-testid="rating-star-unactive"
            number={4}
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            size={15}
          />
          <RatingStar
            active={false}
            color="#ccc"
            data-testid="rating-star-unactive"
            number={5}
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            size={15}
          />
        </div>,
      ],
      "className": "rating-stars",
      "style": Object {
        "height": 15,
        "lineHeight": "15px",
        "width": 75,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <RatingStar
              active={true}
              color="#00adb5"
              data-testid="rating-star-active"
              number={1}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />,
            <RatingStar
              active={true}
              color="#00adb5"
              data-testid="rating-star-active"
              number={2}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />,
          ],
          "className": "rating-stars__active",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "active": true,
              "color": "#00adb5",
              "data-testid": "rating-star-active",
              "number": 1,
              "onClick": [Function],
              "onMouseEnter": [Function],
              "onMouseLeave": [Function],
              "size": 15,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "function",
            "props": Object {
              "active": true,
              "color": "#00adb5",
              "data-testid": "rating-star-active",
              "number": 2,
              "onClick": [Function],
              "onMouseEnter": [Function],
              "onMouseLeave": [Function],
              "size": 15,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <RatingStar
              active={false}
              color="#ccc"
              data-testid="rating-star-unactive"
              number={3}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />,
            <RatingStar
              active={false}
              color="#ccc"
              data-testid="rating-star-unactive"
              number={4}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />,
            <RatingStar
              active={false}
              color="#ccc"
              data-testid="rating-star-unactive"
              number={5}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />,
          ],
          "className": "rating-stars__unactive",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "function",
            "props": Object {
              "active": false,
              "color": "#ccc",
              "data-testid": "rating-star-unactive",
              "number": 3,
              "onClick": [Function],
              "onMouseEnter": [Function],
              "onMouseLeave": [Function],
              "size": 15,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "function",
            "props": Object {
              "active": false,
              "color": "#ccc",
              "data-testid": "rating-star-unactive",
              "number": 4,
              "onClick": [Function],
              "onMouseEnter": [Function],
              "onMouseLeave": [Function],
              "size": 15,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "5",
            "nodeType": "function",
            "props": Object {
              "active": false,
              "color": "#ccc",
              "data-testid": "rating-star-unactive",
              "number": 5,
              "onClick": [Function],
              "onMouseEnter": [Function],
              "onMouseLeave": [Function],
              "size": 15,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="rating-stars__active"
          >
            <RatingStar
              active={true}
              color="#00adb5"
              data-testid="rating-star-active"
              number={1}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />
            <RatingStar
              active={true}
              color="#00adb5"
              data-testid="rating-star-active"
              number={2}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />
          </div>,
          <div
            className="rating-stars__unactive"
          >
            <RatingStar
              active={false}
              color="#ccc"
              data-testid="rating-star-unactive"
              number={3}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />
            <RatingStar
              active={false}
              color="#ccc"
              data-testid="rating-star-unactive"
              number={4}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />
            <RatingStar
              active={false}
              color="#ccc"
              data-testid="rating-star-unactive"
              number={5}
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              size={15}
            />
          </div>,
        ],
        "className": "rating-stars",
        "style": Object {
          "height": 15,
          "lineHeight": "15px",
          "width": 75,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <RatingStar
                active={true}
                color="#00adb5"
                data-testid="rating-star-active"
                number={1}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                size={15}
              />,
              <RatingStar
                active={true}
                color="#00adb5"
                data-testid="rating-star-active"
                number={2}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                size={15}
              />,
            ],
            "className": "rating-stars__active",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "active": true,
                "color": "#00adb5",
                "data-testid": "rating-star-active",
                "number": 1,
                "onClick": [Function],
                "onMouseEnter": [Function],
                "onMouseLeave": [Function],
                "size": 15,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "function",
              "props": Object {
                "active": true,
                "color": "#00adb5",
                "data-testid": "rating-star-active",
                "number": 2,
                "onClick": [Function],
                "onMouseEnter": [Function],
                "onMouseLeave": [Function],
                "size": 15,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <RatingStar
                active={false}
                color="#ccc"
                data-testid="rating-star-unactive"
                number={3}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                size={15}
              />,
              <RatingStar
                active={false}
                color="#ccc"
                data-testid="rating-star-unactive"
                number={4}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                size={15}
              />,
              <RatingStar
                active={false}
                color="#ccc"
                data-testid="rating-star-unactive"
                number={5}
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                size={15}
              />,
            ],
            "className": "rating-stars__unactive",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "function",
              "props": Object {
                "active": false,
                "color": "#ccc",
                "data-testid": "rating-star-unactive",
                "number": 3,
                "onClick": [Function],
                "onMouseEnter": [Function],
                "onMouseLeave": [Function],
                "size": 15,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "function",
              "props": Object {
                "active": false,
                "color": "#ccc",
                "data-testid": "rating-star-unactive",
                "number": 4,
                "onClick": [Function],
                "onMouseEnter": [Function],
                "onMouseLeave": [Function],
                "size": 15,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "function",
              "props": Object {
                "active": false,
                "color": "#ccc",
                "data-testid": "rating-star-unactive",
                "number": 5,
                "onClick": [Function],
                "onMouseEnter": [Function],
                "onMouseLeave": [Function],
                "size": 15,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
